antlr_ast/ast.py

```diff
@@ -283,3 +175,529 @@ def reportContextSensitivity(
+    # - a string: uses AstNode._from_fields as visitor
+    # - a tuple ('node_name', 'ast_node_class_method_name'): uses ast_node_class_method_name as visitor
+    # subclasses use _bind_to_visitor to create visit methods for the nodes in _rules on the ParseTreeVisitor
+    # using this information
+    _rules = []
+
+    _priority = 1
+
+    _strict = True
+
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls, *args, **kwargs)
+        # necessary because AST implements this field
+        instance._fields = cls._fields
+        return instance
+
+    def __init__(self, node: BaseNode, fields=None):
+        # TODO: keep reference to node?
+        # TODO: **fields? (easier notation, but hard to name future arguments
+        super().__init__(
+            node.children, node._field_references, node._label_references, node._ctx
+        )
+
+        fields = fields or {}
+        for field, value in fields.items():
+            if field not in self._fields:
+                warnings.warn("Key not in fields: {}".format(field))
+            setattr(self, field, value)
+
+    @classmethod
+    def from_spec(cls, node):
+        # TODO: no fields_spec argument as before
+        field_dict = {}
+        for field_spec in cls._fields_spec:
+            name, path = parse_field_spec(field_spec)
+
+            # _fields_spec can contain field multiple times
+            # e.g. x=a and x=b
+            if field_dict.get(name):
+                # or / elif behaviour
+                continue
+
+            # get node -----
+            field_dict[name] = cls.get_path(node, path)
+        return cls(node, field_dict)
+
+    @classmethod
+    def get_path(cls, node, path):
+        # TODO: can be defined on FieldNode too
+        result = node
+        for i in range(len(path)):
+            result = getattr(result, path[i], None)
+            if result is None:
+                break
+
+        return result
+
+    @classmethod
+    def bind_to_transformer(cls, transformer_cls, default_transform_method="from_spec"):
+        for rule in cls._rules:
+            if isinstance(rule, str):
+                cls_method = default_transform_method
+            else:
+                rule, cls_method = rule[:2]
+            transformer_method = cls.get_transformer(cls_method)
+            bind_to_transformer(transformer_cls, rule, transformer_method)
+
+    @classmethod
+    def get_transformer(cls, method_name):
+        """Get method to bind to visitor"""
+        visit_node = getattr(cls, method_name)
+        assert callable(visit_node)
+
+        def method(self, node):
+            kwargs = {}
+            if inspect.signature(visit_node).parameters.get("helper"):
+                kwargs["helper"] = self
+            return visit_node(node, **kwargs)
+
+        return method
+
+
+# TODO: test: if 'visit' in method, it has to be as 'visit_'
+class AliasVisitor(NodeTransformer):
```

Comment:
Would call AliasNodeTransformer, so it is very clear what this is doing. Is there a reason that this takes BaseTransformer, and then does a lot of work in `__getattr__`, rather than the AliasVisitor methods being constructed up front? (e.g. something composing I guess AliasVisitor and BaseTransformer). It looks like BaseTransformer is being used on for `getattr(self.transformer_visitor, item)`, so a class that is SomeComposedTransformer(BaseTransformer, AliasVisitor) should work okay. (Could be very wrong)

see comment below

Reply:
It's actually a BaseNodeTransformer, creating AliasNodes.